### PR TITLE
Add validation to the plugin settings

### LINKED ISSUE #4446: Custom branding: add validation to the plugin's settings
## Description

We'll add validation to the settings of the plugin, so no invalid values can be set in the `Settings / Configuration` section.

## Objective
- Display errors when the input is invalid.
- Highlight the settings which contain errors.
- Disable configuration saving when there are errors.

## Tasks
- Frontend
    - [x] Add validation to each type of input component when its value changes.
    - [x] Highlight the errors of each input.
      - [x] Display the cause of error.
      - [x] Add warning icon to the setting with the errors.
    - [x] Disable the save button when there are errors due to invalid values.
    - [x] Add unit tests
    - [x] Add unit tests for `InputForm` of each setting.
- Backend
  - [x] Validate route and body parameters of the following endpoints:
    - [x] `PUT /utils/configuration`
    - [x] `PUT /utils/configuration/files/{key}`
  - [x] Tests
    
**Depends on**: #4443

### PR DESCRIPTION
**Pull request by Desvelao**
# Description

This PR:
- Add:
  - Plugin setting validations:
    - Create reusable methods to compose the validation of each setting
    - Frontend `Settings/Configuration` form:
      - Display the validation error
      - Highlight the setting with invalid input
      - Display the count of settings with an invalid value
      - Disable the save button when there is some invalid value in some setting changed
    - Backend:
      - `PUT /utils/configuration`
        - All settings are validated.
- Test:
  - `useForm` hook. Validate values.
  - `InputFrom` component renders the validation errors. Generic.
  - validate the value for each setting when updating from API endpoint `PUT /utils/configuration`
      
Closes #4446

# Screenshots

- Input's validation, invalid settings highlight, error count display, and disabled of the save button.
image
 
- Warning icon added to the invalid settings, reason displayed on mouseover.
image

# Tests

## Manual tests
**Scenario**: Set an invalid value for the setting in `Settings/Configuration` form
  **When** the user changes the input value to something invalid. This depends on each setting validation.
  **Then** an error should appear below the input form.
  **And** the setting should be red colored and highlighted as invalid.
  **And** the bottom bar should display the count settings with invalid values.
  **And** the save button should be disabled.
  
**Scenario**: Set 2 or more invalid values for the settings in `Settings/Configuration` form
  **When** the user changes the input value of 2 or more settings to invalid values. This depends on each setting validation.
  **Then** an error should appear below the input form for each setting.
  **And** the settings should be red colored and highlighted as invalid.
  **And** the bottom bar should display the count settings with invalid values.
  **And** the save button should be disabled.
  
**Scenario**: Set a valid value for the setting in `Settings/Configuration` form
  **When** the user changes the input value to a valid value. This depends on each setting validation.
  **Then** it should not appear as a validation error below the input.
  
**Scenario**: Send invalid values to update the plugin configuration through the plugin API endpoint
  **When** the user sends (using any rest client) a request to `PUT /utils/configuration` with one or more settings, where at least one of them has an invalid value.
  **Then** it should reply with a bad request displaying the validation error.
  
## Automatic tests
- `server/routes/wazuh-utils/wazuh-utils.test.ts`
- `public/components/common/form/hooks.test.tsx`
- `public/components/common/form/index.test.tsx`
- `common/plugin-settings.test.ts`

**Run all PR tests**
```shell
yarn test:jest server/routes/wazuh-utils/wazuh-utils.test.ts public/components/common/form/hooks.test.tsx public/components/common/form/index.test.tsx common/plugin-settings.test.ts
```

### CODE REVIEW

**Comment by Desvelao** [comment]
# Changes

- Fix some settings validation errors
- Update the development branch with the last changes of the base branch

**Comment by Desvelao** [comment]
# Update
- Update the development branch with the last changes of the base branch.

**Comment by Desvelao** [comment]
# Update

- Update the development branch with the last changes of the base branch.

**Comment by Desvelao** [comment]
# Update

- Add tests to validate each setting in the endpoint to update the plugin configuration on the backend.
- Add tests related to validation for `useForm` hook  and `InputForm` component

**Comment by Desvelao** [comment]
# Suggested enhancement

- The name of indices doesn't support any characters. For Kibana 7.10.2, they could be: https://github.com/elastic/elasticsearch/blob/v7.10.2/docs/reference/indices/create-index.asciidoc
> affected settings:  `alerts.sample.prefix` , `cron.prefix`, `cron.statistics.index.name`
- Index patterns name. Review if there are unsupported characters. They could be the same as the indices.
> affected settings: `pattern`,  `wazuh.monitoring.pattern`, `ip.ignore`

**Comment by Desvelao** [comment]
# Update
- Enhance the validation of some plugin settings related to indices and index patterns.
- Add tests for these validations.
- Add tests for the input valdiation.

**Comment by Desvelao** [comment]
# Update
- Update the development branch with the base branch

File: common/services/settings-validate.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
+// Utils
```
Comment (AlexRuiz7) [inline common/services/settings-validate.ts:1]: We need to discuss this service

- The name does not make much sense in English. `settings-validator` or `settings-validation` makes sense, but `settings-validate` doesn't.
- A function or method which responsibility is to check or validate something, usually returns a boolean. Most of the functions defined here not only return not boolean values, but return `undefined`. This is not only weird, it's also dangerous and prone to bugs. I understand that these strings are used as messages to be shown in the input for feedback. In this case, the following data structure could be used:

```ts
interface SettingValidation {
  isValid: boolean,
  reason: string
}

// ------------ //

function isString(value: any): SettingValidation {};

// ------------ //

const validationResult = isString(13);
if (validationResult.isValid) {
  input.validationError = validationResult.reason;
}
```

Additionally, this service could be a class:

```ts

class SettingsValidator {

  static isString() {};
  static isEmptyString() {};
  static stringHasMultipleLines() {};
  ...
  static isBoolean() {};
}

// ------------ //

const validationResult = SettingValidator.isString(13);
```

Finally, function declarations are recommended over function expression, following our **[Styleguide](https://github.com/wazuh/wazuh-kibana-app/blob/master/STYLEGUIDE.md#function-definitions)**.

File: common/constants.ts [hunk 2]
```diff
@@ -11,6 +11,8 @@
  */
 import path from 'path';
 import { version } from '../package.json';
+import { validate as validateNodeCronInterval } from 'node-cron';
+import { composeValidate, validateBooleanIs, validateLiteral, validateNumber, validateStringNoInvalidCharacters, validateStringNoEmpty, validateStringNoLiteral, validateStringNoSpaces, validateStringNoStartWith, validateJSON, validateObjectArray, validateStringIs } from './services/settings-validate';
```
Comment (AlexRuiz7) [inline common/constants.ts:15]: Following my previous comment, a single import would be needed.

```ts
import SettingsValidator from './services/settings-validator';
```

**Review by yenienserrano — DISMISSED** [review]
LGTM

File: common/services/settings-validate.ts (hunk 1 above)
Comment (Desvelao) [reply common/services/settings-validate.ts:1]: > We need to discuss this service
> 
> * The name does not make much sense in English. `settings-validator` or `settings-validation` makes sense, but `settings-validate` doesn't.
> * A function or method which responsibility is to check or validate something, usually returns a boolean. Most of the functions defined here not only return not boolean values, but return `undefined`. This is not only weird, it's also dangerous and prone to bugs. I understand that these strings are used as messages to be shown in the input for feedback. In this case, the following data structure could be used:
> 
> ```ts
> interface SettingValidation {
>   isValid: boolean,
>   reason: string
> }
> 
> // ------------ //
> 
> function isString(value: any): SettingValidation {};
> 
> // ------------ //
> 
> const validationResult = isString(13);
> if (validationResult.isValid) {
>   input.validationError = validationResult.reason;
> }
> ```

> Finally, function declarations are recommended over function expression, following our **[Styleguide](https://github.com/wazuh/wazuh-kibana-app/blob/master/STYLEGUIDE.md#function-definitions)**.

The validation functions are used in the server and frontend.

Their definition (input/output) is based on the `validate` parameter of `@kbn/config-schema` package. They return a `string` or `undefined`. The current definition allows us to use it directly as the `validate` parameter without wrappers. More information at https://github.com/elastic/kibana/tree/v7.10.2/packages/kbn-config-schema#custom-validation. This means that returning an object as it is suggested:
```ts
interface SettingValidation {
  isValid: boolean,
  reason: string
}
```
will not be compatible if we use that function as the `validate` property so we will need to wrap it to manage the output returning a `string` or `undefined`.

I think the approach for the validation is simple depending on the output type:
- `string`: the reason for the validation error.
- `undefined`: no validation error.

In the frontend, these functions are used to validate the input in the form of `Settings/Configuration`. If these functions return a string, then we can:
- display the field is invalid
- display the reason for the validation error

One value is used for 2 things.

> Additionally, this service could be a class:
> 
> ```ts
> class SettingsValidator {
> 
>   static isString() {};
>   static isEmptyString() {};
>   static stringHasMultipleLines() {};
>   ...
>   static isBoolean() {};
> }
> 
> // ------------ //
> 
> const validationResult = SettingValidator.isString(13);
> ```

We should discuss it and define it in the style guide. Functionally and in this case, functions are the same as classes with static methods.

**Comment by Desvelao** [comment]
# Update
- Fix a bug in the `useForm` hook
- Refactor the definition of the setting validation functions to a class with static methods

**Comment by Desvelao** [comment]
# Update
- Enhance the validation of the settings of type `number` checking if these are integers.
image

**Review by AlexRuiz7 — DISMISSED** [review]
LGTM

**Review by AlexRuiz7 — APPROVED** [review]
LGTM

**Review by AlexRuiz7 — APPROVED** [review]
LGTM

**Review by yenienserrano — APPROVED** [review]
LGTM

**Comment by Mayons95** [comment]
TR: :heavy_check_mark: 
<details>

[Screencast from 26-10-22 10:02:57.webm](https://user-images.githubusercontent.com/104914131/198033485-70933096-0ec1-4d7b-8c89-50772bb2d841.webm)

[Screencast from 26-10-22 09:37:23.webm](https://user-images.githubusercontent.com/104914131/198033598-8a763615-a913-4de2-ae0d-19e0bb99d98d.webm)


</details>

**Review by Mayons95 — DISMISSED** [review]
LGTM

**Review by Mayons95 — APPROVED** [review]
TR: :heavy_check_mark: 
LGTM

**Review by yenienserrano — APPROVED** [review]
LGTM

**Pull request merged**